### PR TITLE
Ge/lookup by name

### PR DESCRIPTION
**Pull request by gegnew**
This adds a `createLookup` function that allows specifying a resource by it's name, i.e.:
```
lookup = createLookup("5e1f66a06f5f3f0759b479c9")
lookup("gates", "test_gate")
```

It leverages the existing internal `lookupByName` function on a per-experiment basis, so after one run (to populate the cache) it should be about as fast as querying by id.

To test the `createLookup` function (and thereby implicitly test the `lookupByName` function) I had to add a line to the `lookupByName` function that ensures all  responses are coerced into a `dataframe`, otherwise the tests will fail (since `httmock` responses are serialized to a list instead of to a dataframe).

### CODE REVIEW

File: R/util.R [hunk 1]
```diff
@@ -142,3 +142,75 @@ lookupByName = function(listpath, name, prop = "name") {
...
+
+#' Find a Resource By Name
+#'
+#' Searches for any resource by name. Accepts imperfect resource
+#' specification, (i.e. "gate" for "gates").
+#'
+#' Because two resources in different experiments could have the
+#' same name, this clears the cache of ids and names each time it runs.
+#' It is recommended that you use the regular lookup functions
+#' for performant code.
```
Comment (zbjornson) [inline R/util.R]: This function should take the `experimentId`, otherwise it'll be an impossibly large search space. That avoids this issue.

Comment (gegnew) [reply R/util.R]: It seemed from the issue description that we wanted to be able to query _any_ name of any resource, which requires checking all experiments; do you want it to work like `findByName(experimentId, resourceType, name)` instead?

File: R/util.R [hunk 2]
```diff
@@ -142,3 +142,75 @@ lookupByName = function(listpath, name, prop = "name") {
...
+
+    result = try(getter(id, byName(name)), silent=TRUE)
+    if(class(result) != "try-error") {
+      results = rbind(results, result)
+    }
+  }
+  return(results)
+}
+
+}
```
[the remark this answers is not in the document]
Comment (gegnew) [reply R/util.R]: Yup, good catch

File: NAMESPACE [hunk 3]
```diff
@@ -19,6 +20,7 @@ export(deleteExperiment)
 export(deleteFcsFile)
 export(deleteGates)
 export(deletePopulationAuto)
+export(findByName)
```
Comment (zbjornson) [inline NAMESPACE]: This doesn't exist anymore, right?

File: R/getGate.R [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#' Get a gate
+#'
+#' Retrieves the list of gates in an experiment.
```
Comment (zbjornson) [inline R/getGate.R]: s/the list of gates/the gate/

File: R/getGate.R [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+#'
+#' @param experimentId ID of experiment.
+#' @param params Optional query parameters.
+#' @export
+#' @examples
+#' \dontrun{
+#' getGate(experimentId, gateId)
+#' getGate(experimentId, gateId, params = list("limit" = "5"))
```
Comment (zbjornson) [inline R/getGate.R]: You can remove the `params` arg (the API takes non).

File: R/util.R [hunk 6]
```diff
@@ -142,3 +142,35 @@ lookupByName = function(listpath, name, prop = "name") {
   assign(key, val, envir = byNameHash)
   val
 }
+
+#' Find a Resource By Name
+#'
+#' Allows createion of a by-experiment lookup object, which can
```
Comment (zbjornson) [inline R/util.R]: typo

File: R/util.R [hunk 7]
```diff
@@ -142,3 +142,35 @@ lookupByName = function(listpath, name, prop = "name") {
...
+#' then be used to find resources by name instead of id. Utilizes
+#' the internal lookupByName cache to improve performance.
```
Comment (zbjornson) [inline R/util.R]: This is an implementation detail we don't need to reveal to users. The next sentence is sufficient.

File: R/util.R [hunk 8]
```diff
@@ -142,3 +142,35 @@ lookupByName = function(listpath, name, prop = "name") {
...
+#' The resource's id is cached for the duration of the R
+#' session. Resources such as gates that exist within an experiment
+#' are cached within the experiment's scope.
+#'
+#' \code{
+#' lookup = createLookup("5e1f66a06f5f3f0759b479c9")
+#' lookup("gates", "test_gate")
+#'
+#' @param experimentId Experiment to create lookup for.
+#' Internal function params:
```
Comment (zbjornson) [inline R/util.R]: Maybe "returned function params" instead of "internal"?

File: R/util.R [hunk 9]
```diff
@@ -142,3 +142,35 @@ lookupByName = function(listpath, name, prop = "name") {
...
+#' @param resource Type of resource.
```
Comment (zbjornson) [inline R/util.R]: I'd list the available options and/or say that they have to be pluralized.

Probably also add a validator for the `resource` argument -- `resource %in% c("gates", ...)`

File: R/getGate.R (hunk 5 above)
Comment (gegnew) [reply R/getGate.R]: right, only for _list_ routes

**Pull request merged**